// @applitools/eyes/dist/output/ApiUrls.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiUrlsData = void 0;
const utils = __importStar(require("@applitools/utils"));
class ApiUrlsData {
    /** @internal */
    constructor(urls) {
        this._urls = urls instanceof ApiUrlsData ? urls.toJSON() : urls;
    }
    get baselineImage() {
        return this._urls.baselineImage;
    }
    getBaselineImage() {
        return this.baselineImage;
    }
    /** @deprecated */
    setBaselineImage(setBaselineImage) {
        this._urls.baselineImage = setBaselineImage;
    }
    get currentImage() {
        return this._urls.currentImage;
    }
    getCurrentImage() {
        return this.currentImage;
    }
    /** @deprecated */
    setCurrentImage(currentImage) {
        this._urls.currentImage = currentImage;
    }
    get checkpointImage() {
        return this._urls.checkpointImage;
    }
    getCheckpointImage() {
        return this.checkpointImage;
    }
    /** @deprecated */
    setCheckpointImage(checkpointImage) {
        this._urls.checkpointImage = checkpointImage;
    }
    get checkpointImageThumbnail() {
        return this._urls.checkpointImageThumbnail;
    }
    getCheckpointImageThumbnail() {
        return this.checkpointImageThumbnail;
    }
    /** @deprecated */
    setCheckpointImageThumbnail(checkpointImageThumbnail) {
        this._urls.checkpointImageThumbnail = checkpointImageThumbnail;
    }
    get diffImage() {
        return this._urls.diffImage;
    }
    getDiffImage() {
        return this.diffImage;
    }
    /** @deprecated */
    setDiffImage(diffImage) {
        this._urls.diffImage = diffImage;
    }
    /** @internal */
    toObject() {
        return this._urls;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._urls);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.ApiUrlsData = ApiUrlsData;
