// @applitools/eyes/dist/output/StepInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepInfoData = void 0;
const utils = __importStar(require("@applitools/utils"));
const AppUrls_1 = require("./AppUrls");
const ApiUrls_1 = require("./ApiUrls");
class StepInfoData {
    /** @internal */
    constructor(info) {
        this._info = info instanceof StepInfoData ? info.toJSON() : info;
    }
    get name() {
        return this._info.name;
    }
    getName() {
        return this.name;
    }
    /** @deprecated */
    setName(value) {
        this._info.name = value;
    }
    get isDifferent() {
        return this._info.isDifferent;
    }
    getIsDifferent() {
        return this.isDifferent;
    }
    /** @deprecated */
    setIsDifferent(value) {
        this._info.isDifferent = value;
    }
    get hasBaselineImage() {
        return this._info.hasBaselineImage;
    }
    getHasBaselineImage() {
        return this.hasBaselineImage;
    }
    /** @deprecated */
    setHasBaselineImage(value) {
        this._info.hasBaselineImage = value;
    }
    get hasCurrentImage() {
        return this._info.hasCurrentImage;
    }
    getHasCurrentImage() {
        return this.hasCurrentImage;
    }
    /** @deprecated */
    setHasCurrentImage(hasCurrentImage) {
        this._info.hasCurrentImage = hasCurrentImage;
    }
    get appUrls() {
        return this._info.appUrls;
    }
    getAppUrls() {
        return new AppUrls_1.AppUrlsData(this.appUrls);
    }
    /** @deprecated */
    setAppUrls(appUrls) {
        this._info.appUrls = appUrls;
    }
    get apiUrls() {
        return this._info.apiUrls;
    }
    getApiUrls() {
        return new ApiUrls_1.ApiUrlsData(this.apiUrls);
    }
    /** @deprecated */
    setApiUrls(apiUrls) {
        this._info.apiUrls = apiUrls;
    }
    get renderId() {
        return this._info.renderId;
    }
    getRenderId() {
        return this.renderId;
    }
    /** @deprecated */
    setRenderId(renderId) {
        this._info.renderId = renderId;
    }
    /** @internal */
    toObject() {
        return this._info;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._info);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.StepInfoData = StepInfoData;
