// @applitools/eyes/dist/enums/AndroidDeviceName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AndroidDeviceNameEnum = void 0;
var AndroidDeviceNameEnum;
(function (AndroidDeviceNameEnum) {
    AndroidDeviceNameEnum["Pixel_3_XL"] = "Pixel 3 XL";
    AndroidDeviceNameEnum["Pixel_4"] = "Pixel 4";
    AndroidDeviceNameEnum["Pixel_4_XL"] = "Pixel 4 XL";
    AndroidDeviceNameEnum["Pixel_5"] = "Pixel 5";
    AndroidDeviceNameEnum["Pixel_6"] = "Pixel 6";
    AndroidDeviceNameEnum["Galaxy_S10"] = "Galaxy S10";
    AndroidDeviceNameEnum["Galaxy_S10_Plus"] = "Galaxy S10 Plus";
    AndroidDeviceNameEnum["Galaxy_Note_10"] = "Galaxy Note 10";
    AndroidDeviceNameEnum["Galaxy_Note_10_Plus"] = "Galaxy Note 10 Plus";
    AndroidDeviceNameEnum["Galaxy_S20"] = "Galaxy S20";
    AndroidDeviceNameEnum["Galaxy_S20_Plus"] = "Galaxy S20 Plus";
    AndroidDeviceNameEnum["Galaxy_S21"] = "Galaxy S21";
    AndroidDeviceNameEnum["Galaxy_S21_Plus"] = "Galaxy S21 Plus";
    AndroidDeviceNameEnum["Galaxy_S21_Ultra"] = "Galaxy S21 Ultra";
    AndroidDeviceNameEnum["Galaxy_S22"] = "Galaxy S22";
    AndroidDeviceNameEnum["Galaxy_S22_Plus"] = "Galaxy S22 Plus";
    AndroidDeviceNameEnum["Galaxy_Tab_S7"] = "Galaxy Tab S7";
    AndroidDeviceNameEnum["Galaxy_Tab_S8"] = "Galaxy Tab S8";
    AndroidDeviceNameEnum["Xiaomi_Redmi_Note_10_JE"] = "Xiaomi Redmi Note 10 JE";
    AndroidDeviceNameEnum["Xiaomi_Redmi_Note_11"] = "Xiaomi Redmi Note 11";
    AndroidDeviceNameEnum["Xiaomi_Redmi_Note_11_Pro"] = "Xiaomi Redmi Note 11 Pro";
    AndroidDeviceNameEnum["Sony_Xperia_1_II"] = "Sony Xperia 1 II";
    AndroidDeviceNameEnum["Sony_Xperia_10_II"] = "Sony Xperia 10 II";
    AndroidDeviceNameEnum["Sony_Xperia_Ace_II"] = "Sony Xperia Ace II";
    AndroidDeviceNameEnum["Huawei_P30_Lite"] = "Huawei P30 Lite";
})(AndroidDeviceNameEnum || (exports.AndroidDeviceNameEnum = AndroidDeviceNameEnum = {}));
