// @applitools/eyes/dist/enums/BrowserType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserTypeEnum = void 0;
var BrowserTypeEnum;
(function (BrowserTypeEnum) {
    BrowserTypeEnum["CHROME"] = "chrome";
    BrowserTypeEnum["CHROME_ONE_VERSION_BACK"] = "chrome-one-version-back";
    BrowserTypeEnum["CHROME_TWO_VERSIONS_BACK"] = "chrome-two-versions-back";
    BrowserTypeEnum["FIREFOX"] = "firefox";
    BrowserTypeEnum["FIREFOX_ONE_VERSION_BACK"] = "firefox-one-version-back";
    BrowserTypeEnum["FIREFOX_TWO_VERSIONS_BACK"] = "firefox-two-versions-back";
    BrowserTypeEnum["IE_11"] = "ie";
    BrowserTypeEnum["IE_10"] = "ie10";
    BrowserTypeEnum["EDGE"] = "edge";
    BrowserTypeEnum["EDGE_CHROMIUM"] = "edgechromium";
    BrowserTypeEnum["EDGE_LEGACY"] = "edgelegacy";
    BrowserTypeEnum["EDGE_CHROMIUM_ONE_VERSION_BACK"] = "edgechromium-one-version-back";
    BrowserTypeEnum["EDGE_CHROMIUM_TWO_VERSIONS_BACK"] = "edgechromium-two-versions-back";
    BrowserTypeEnum["SAFARI"] = "safari";
    BrowserTypeEnum["SAFARI_EARLY_ACCESS"] = "safari-earlyaccess";
    BrowserTypeEnum["SAFARI_ONE_VERSION_BACK"] = "safari-one-version-back";
    BrowserTypeEnum["SAFARI_TWO_VERSIONS_BACK"] = "safari-two-versions-back";
})(BrowserTypeEnum || (exports.BrowserTypeEnum = BrowserTypeEnum = {}));
