// @applitools/eyes/dist/enums/DeviceName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceNameEnum = void 0;
var DeviceNameEnum;
(function (DeviceNameEnum) {
    DeviceNameEnum["Blackberry_PlayBook"] = "Blackberry PlayBook";
    DeviceNameEnum["BlackBerry_Z30"] = "BlackBerry Z30";
    DeviceNameEnum["Galaxy_A5"] = "Galaxy A5";
    DeviceNameEnum["Galaxy_Note_10"] = "Galaxy Note 10";
    DeviceNameEnum["Galaxy_Note_10_Plus"] = "Galaxy Note 10 Plus";
    DeviceNameEnum["Galaxy_Note_2"] = "Galaxy Note 2";
    DeviceNameEnum["Galaxy_Note_3"] = "Galaxy Note 3";
    DeviceNameEnum["Galaxy_Note_4"] = "Galaxy Note 4";
    DeviceNameEnum["Galaxy_Note_8"] = "Galaxy Note 8";
    DeviceNameEnum["Galaxy_Note_9"] = "Galaxy Note 9";
    DeviceNameEnum["Galaxy_S3"] = "Galaxy S3";
    DeviceNameEnum["Galaxy_S5"] = "Galaxy S5";
    DeviceNameEnum["Galaxy_S8"] = "Galaxy S8";
    DeviceNameEnum["Galaxy_S8_Plus"] = "Galaxy S8 Plus";
    DeviceNameEnum["Galaxy_S9"] = "Galaxy S9";
    DeviceNameEnum["Galaxy_S9_Plus"] = "Galaxy S9 Plus";
    DeviceNameEnum["Galaxy_S10"] = "Galaxy S10";
    DeviceNameEnum["Galaxy_S10_Plus"] = "Galaxy S10 Plus";
    DeviceNameEnum["Galaxy_S20"] = "Galaxy S20";
    DeviceNameEnum["Galaxy_S22"] = "Galaxy S22";
    DeviceNameEnum["Galaxy_Tab_S7"] = "Galaxy Tab S7";
    DeviceNameEnum["iPad"] = "iPad";
    DeviceNameEnum["iPad_6th_Gen"] = "iPad 6th Gen";
    DeviceNameEnum["iPad_7th_Gen"] = "iPad 7th Gen";
    DeviceNameEnum["iPad_Air_2"] = "iPad Air 2";
    DeviceNameEnum["iPad_Mini"] = "iPad Mini";
    DeviceNameEnum["iPad_Pro"] = "iPad Pro";
    DeviceNameEnum["iPhone_11"] = "iPhone 11";
    DeviceNameEnum["iPhone_11_Pro"] = "iPhone 11 Pro";
    DeviceNameEnum["iPhone_11_Pro_Max"] = "iPhone 11 Pro Max";
    DeviceNameEnum["iPhone_4"] = "iPhone 4";
    DeviceNameEnum["iPhone_5SE"] = "iPhone 5/SE";
    DeviceNameEnum["iPhone_6_7_8"] = "iPhone 6/7/8";
    DeviceNameEnum["iPhone_6_7_8_Plus"] = "iPhone 6/7/8 Plus";
    DeviceNameEnum["iPhone_X"] = "iPhone X";
    DeviceNameEnum["iPhone_XR"] = "iPhone XR";
    DeviceNameEnum["iPhone_XS"] = "iPhone XS";
    DeviceNameEnum["iPhone_XS_Max"] = "iPhone XS Max";
    DeviceNameEnum["Kindle_Fire_HDX"] = "Kindle Fire HDX";
    DeviceNameEnum["Laptop_with_HiDPI_screen"] = "Laptop with HiDPI screen";
    DeviceNameEnum["Laptop_with_MDPI_screen"] = "Laptop with MDPI screen";
    DeviceNameEnum["Laptop_with_touch"] = "Laptop with touch";
    DeviceNameEnum["LG_G6"] = "LG G6";
    DeviceNameEnum["LG_Optimus_L70"] = "LG Optimus L70";
    DeviceNameEnum["Microsoft_Lumia_550"] = "Microsoft Lumia 550";
    DeviceNameEnum["Microsoft_Lumia_950"] = "Microsoft Lumia 950";
    DeviceNameEnum["Nexus_10"] = "Nexus 10";
    DeviceNameEnum["Nexus_4"] = "Nexus 4";
    DeviceNameEnum["Nexus_5"] = "Nexus 5";
    DeviceNameEnum["Nexus_5X"] = "Nexus 5X";
    DeviceNameEnum["Nexus_6"] = "Nexus 6";
    DeviceNameEnum["Nexus_6P"] = "Nexus 6P";
    DeviceNameEnum["Nexus_7"] = "Nexus 7";
    DeviceNameEnum["Nokia_Lumia_520"] = "Nokia Lumia 520";
    DeviceNameEnum["Nokia_N9"] = "Nokia N9";
    DeviceNameEnum["OnePlus_7T"] = "OnePlus 7T";
    DeviceNameEnum["OnePlus_7T_Pro"] = "OnePlus 7T Pro";
    // OnePlus_8 = 'OnePlus 8',
    // OnePlus_8_Pro = 'OnePlus 8 Pro',
    DeviceNameEnum["Pixel_2"] = "Pixel 2";
    DeviceNameEnum["Pixel_2_XL"] = "Pixel 2 XL";
    DeviceNameEnum["Pixel_3"] = "Pixel 3";
    DeviceNameEnum["Pixel_3_XL"] = "Pixel 3 XL";
    DeviceNameEnum["Pixel_4"] = "Pixel 4";
    DeviceNameEnum["Pixel_4_XL"] = "Pixel 4 XL";
    DeviceNameEnum["Pixel_5"] = "Pixel 5";
    DeviceNameEnum["Sony_Xperia_10_II"] = "Sony Xperia 10 II";
    DeviceNameEnum["Huawei_Mate_50_Pro"] = "Huawei Mate 50 Pro";
    DeviceNameEnum["Huawei_Matepad_11"] = "Huawei Matepad 11";
})(DeviceNameEnum || (exports.DeviceNameEnum = DeviceNameEnum = {}));
