// @applitools/eyes/dist/enums/IosDeviceName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IosDeviceNameEnum = void 0;
var IosDeviceNameEnum;
(function (IosDeviceNameEnum) {
    IosDeviceNameEnum["iPhone_14_Pro_Max"] = "iPhone 14 Pro Max";
    IosDeviceNameEnum["iPhone_14"] = "iPhone 14";
    IosDeviceNameEnum["iPhone_13_Pro_Max"] = "iPhone 13 Pro Max";
    IosDeviceNameEnum["iPhone_13_Pro"] = "iPhone 13 Pro";
    IosDeviceNameEnum["iPhone_13"] = "iPhone 13";
    IosDeviceNameEnum["iPhone_12_Pro_Max"] = "iPhone 12 Pro Max";
    IosDeviceNameEnum["iPhone_12_Pro"] = "iPhone 12 Pro";
    IosDeviceNameEnum["iPhone_12"] = "iPhone 12";
    IosDeviceNameEnum["iPhone_12_mini"] = "iPhone 12 mini";
    IosDeviceNameEnum["iPhone_11_Pro"] = "iPhone 11 Pro";
    IosDeviceNameEnum["iPhone_11_Pro_Max"] = "iPhone 11 Pro Max";
    IosDeviceNameEnum["iPhone_11"] = "iPhone 11";
    IosDeviceNameEnum["iPhone_XR"] = "iPhone XR";
    IosDeviceNameEnum["iPhone_XS"] = "iPhone Xs";
    IosDeviceNameEnum["iPhone_X"] = "iPhone X";
    IosDeviceNameEnum["iPhone_8"] = "iPhone 8";
    IosDeviceNameEnum["iPhone_8_Plus"] = "iPhone 8 Plus";
    IosDeviceNameEnum["iPhone_7"] = "iPhone 7";
    IosDeviceNameEnum["iPhone_SE"] = "iPhone SE (1st generation)";
    IosDeviceNameEnum["iPad_Pro_3"] = "iPad Pro (12.9-inch) (3rd generation)";
    IosDeviceNameEnum["iPad_Pro_4"] = "iPad Pro (11-inch) (4th generation)";
    IosDeviceNameEnum["iPad_7"] = "iPad (7th generation)";
    IosDeviceNameEnum["iPad_9"] = "iPad (9th generation)";
    IosDeviceNameEnum["iPad_Air_2"] = "iPad Air (2nd generation)";
    IosDeviceNameEnum["iPad_Air_4"] = "iPad Air (4th generation)";
})(IosDeviceNameEnum || (exports.IosDeviceNameEnum = IosDeviceNameEnum = {}));
