// @applitools/core/dist/ufg/utils/take-dom-snapshot.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractCrossFrames = exports.deserializeDomSnapshot = exports.takeDomSnapshot = void 0;
const utils = __importStar(require("@applitools/utils"));
const { getProcessPagePoll, getPollResult, getProcessPagePollForIE, getPollResultForIE, } = require('@applitools/dom-snapshot');
async function takeDomSnapshot({ context, settings, logger, }) {
    const driver = context.driver;
    const environment = await driver.getEnvironment();
    const features = await driver.getFeatures();
    const cookies = features.allCookies ? await driver.getCookies().catch(() => []) : [];
    const snapshot = deserializeDomSnapshot({ snapshot: await takeContextDomSnapshot({ context }) });
    snapshot.cookies = cookies;
    return snapshot;
    async function takeContextDomSnapshot({ context }) {
        // logger.log(`taking dom snapshot. ${context._reference ? `context referece: ${JSON.stringify(context._reference)}` : ''}`)
        var _a, _b, _c;
        if (!features.allCookies) {
            cookies.push(...(await context.getCookies()));
        }
        const isLegacyBrowser = environment.isIE || environment.isEdgeLegacy;
        const arg = {
            dontFetchResources: settings === null || settings === void 0 ? void 0 : settings.disableBrowserFetching,
            skipResources: settings === null || settings === void 0 ? void 0 : settings.skipResources,
            removeReverseProxyURLPrefixes: Boolean(process.env.APPLITOOLS_SCRIPT_REMOVE_REVERSE_PROXY_URL_PREFIXES),
            chunkByteLength: (_a = settings === null || settings === void 0 ? void 0 : settings.chunkByteLength) !== null && _a !== void 0 ? _a : (Number(process.env.APPLITOOLS_SCRIPT_RESULT_MAX_BYTE_LENGTH) ||
                (environment.isIOS ? 100000 : 250 * 1024 * 1024)),
            serializeResources: true,
            compressResources: false,
            showLogs: settings === null || settings === void 0 ? void 0 : settings.showLogs,
        };
        const scripts = {
            main: features.canExecuteOnlyFunctionScripts
                ? require('@applitools/dom-snapshot').processPagePoll
                : `return (${isLegacyBrowser ? await getProcessPagePollForIE() : await getProcessPagePoll()}).apply(null, arguments);`,
            poll: features.canExecuteOnlyFunctionScripts
                ? require('@applitools/dom-snapshot').pollResult
                : `return (${isLegacyBrowser ? await getPollResultForIE() : await getPollResult()}).apply(null, arguments);`,
        };
        const snapshot = await context.executePoll(scripts, {
            main: arg,
            poll: arg,
            executionTimeout: (_b = settings === null || settings === void 0 ? void 0 : settings.executionTimeout) !== null && _b !== void 0 ? _b : 5 * 60 * 1000,
            pollTimeout: (_c = settings === null || settings === void 0 ? void 0 : settings.pollTimeout) !== null && _c !== void 0 ? _c : 200,
        });
        const crossFrames = extractCrossFrames({ snapshot, logger });
        for (const { reference, parentSnapshot, cdtNode } of crossFrames) {
            const frameContext = await context
                .context(reference)
                .then(context => context.focus())
                .catch(err => {
                const srcAttr = cdtNode.attributes.find(attr => attr.name === 'src');
                if (srcAttr)
                    srcAttr.value = '';
                logger.log(`could not switch to frame during takeDomSnapshot. Path to frame: ${JSON.stringify(reference)}`, err);
            });
            if (frameContext) {
                const frameSnapshot = await takeContextDomSnapshot({ context: frameContext });
                let url = new URL(frameSnapshot.url);
                if (url.protocol === 'data:')
                    url = new URL(`http://data-url-frame${url.search}`);
                if (!url.searchParams.has('applitools-iframe'))
                    url.searchParams.set('applitools-iframe', utils.general.guid());
                frameSnapshot.url = url.href;
                parentSnapshot.frames.push(frameSnapshot);
                cdtNode.attributes.push({ name: 'data-applitools-src', value: frameSnapshot.url });
            }
        }
        logger.log(`dom snapshot cdt length: ${snapshot.cdt.length}`);
        logger.log(`blobs urls (${snapshot.blobs.length}):`, JSON.stringify(snapshot.blobs.map(({ url }) => url)));
        logger.log(`resource urls (${snapshot.resourceUrls.length}):`, JSON.stringify(snapshot.resourceUrls));
        return snapshot;
    }
}
exports.takeDomSnapshot = takeDomSnapshot;
function deserializeDomSnapshot({ snapshot }) {
    const { blobs, selector: _, crossFrames: __, ...rest } = snapshot;
    const deserializedSnapshot = {
        ...rest,
        resourceContents: blobs.reduce((resourceContents, blob) => {
            if (blob.value === undefined)
                return { ...resourceContents, [blob.url]: blob };
            else
                return { ...resourceContents, [blob.url]: { ...blob, value: Buffer.from(blob.value, 'base64') } };
        }, {}),
        frames: snapshot.frames.map(frameSnapshot => deserializeDomSnapshot({ snapshot: frameSnapshot })),
    };
    return deserializedSnapshot;
}
exports.deserializeDomSnapshot = deserializeDomSnapshot;
function extractCrossFrames({ snapshot, parent = null, logger, }) {
    var _a;
    const crossFrames = [snapshot, ...((_a = snapshot.frames) !== null && _a !== void 0 ? _a : [])].flatMap((snapshot, index) => {
        var _a;
        const crossFrames = ((_a = snapshot.crossFrames) !== null && _a !== void 0 ? _a : []).map(({ selector, index }) => ({
            reference: { reference: { type: 'css', selector }, parent },
            parentSnapshot: snapshot,
            cdtNode: snapshot.cdt[index],
        }));
        return [
            ...crossFrames,
            ...(index > 0
                ? extractCrossFrames({
                    snapshot,
                    parent: { reference: { type: 'css', selector: snapshot.selector }, parent },
                    logger,
                })
                : []),
        ];
    });
    logger.log(`frames paths for ${snapshot.crossFrames}`, crossFrames.map(selector => JSON.stringify(selector)).join(' , '));
    return crossFrames;
}
exports.extractCrossFrames = extractCrossFrames;
