// @applitools/eg-socks5-proxy-server/src/events.ts
export const AUTHENTICATION = 'authenticate'
export const AUTHENTICATION_ERROR = 'authenticateError'
export const CONNECTION_FILTER = 'connectionFilter'
export const HANDSHAKE = 'handshake'
export const PROXY_END = 'proxyEnd'

export const ACCEPT_NEW_REQUEST = 'accept-new-request'
export const CONNECTED_TO_REMOTE_ADDRESS = 'connected-to-remote-address'
export const CONNECTED_TO_REMOTE_PROXY = 'connected-to-remote-proxy'
export const REMOTE_PROXY_HANDSHAKE_COMPLETED = 'remote-proxy-handshake-completed'
export const REMOTE_PROXY_HANDSHAKE_TIMEOUT = 'remote-proxy-handshake-timeout'
export const REMOTE_PROXY_RFC_1928_REPLY_TIMEOUT = 'remote-proxy-rpc-1928-reply-timeout'
export const REMOTE_PROXY_RFC_1928_REPLY = 'remote-proxy-rpc-1928-reply'
export const ORIGIN_SOCKET_ERROR = 'orgin-socket-error'
export const REMOTE_SOCKET_ERROR = 'remote-socket-error'
export const REMOTE_CONNECTION_TIMEOUT_ERROR = 'remote-connection-timeout-error'
export const PRIVATE_ADDRESS_WAS_BLOCKED = 'private-address-was-blocked'
