// @applitools/core-base/dist/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCore = void 0;
const logger_1 = require("@applitools/logger");
const requests_1 = require("./server/requests");
const open_eyes_1 = require("./open-eyes");
const open_functional_session_1 = require("./open-functional-session");
const locate_1 = require("./locate");
const locate_text_1 = require("./locate-text");
const extract_text_1 = require("./extract-text");
const close_batch_1 = require("./close-batch");
const delete_test_1 = require("./delete-test");
function makeCore({ agentId = 'core-base', concurrency, cwd = process.cwd(), logger: defaultLogger, }) {
    const logger = (0, logger_1.makeLogger)({ logger: defaultLogger, format: { label: 'core-base' } });
    logger.log(`Core is initialized in directory ${cwd} for agent ${agentId}`);
    const coreRequests = (0, requests_1.makeCoreRequests)({ agentId, logger });
    return {
        openEyes: (0, open_eyes_1.makeOpenEyes)({ requests: coreRequests, concurrency, cwd, logger }),
        openFunctionalSession: (0, open_functional_session_1.makeOpenFunctionalSession)({ requests: coreRequests, concurrency, cwd, logger }),
        locate: (0, locate_1.makeLocate)({ requests: coreRequests, logger }),
        locateText: (0, locate_text_1.makeLocateText)({ requests: coreRequests, logger }),
        extractText: (0, extract_text_1.makeExtractText)({ requests: coreRequests, logger }),
        closeBatch: (0, close_batch_1.makeCloseBatch)({ requests: coreRequests, logger }),
        deleteTest: (0, delete_test_1.makeDeleteTest)({ requests: coreRequests, logger }),
        getAccountInfo: coreRequests.getAccountInfo,
        logEvent: coreRequests.logEvent,
    };
}
exports.makeCore = makeCore;
